### PR TITLE
[LW-DETR] Fix training

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR fixes the training of LW-DETR. It turned out that the model was not able to overfit a single batch. Hence I asked Codex to investigate this. 

It turns out there were 3 bugs:

1. A logits calibration gap, caused by a training-time bug: the encoder output box coordinates are incorrectly detached in the HF model, cutting off gradient flow from the encoder output box loss to the two-stage bbox prediction heads.
2. Train-time group-DETR self-attention in HF used a reshape based on pre-split `(batch, seq_len)` instead of post-split shape.
3. (nit) Converted HF configs were using `dropout=0.1`, while original LW-DETR training config uses `dropout=0.0`.

After this, I was able to fine-tune LW-DETR properly on a sample dataset (https://huggingface.co/datasets/nielsr/tray-cart-detection) with parity with respect to the original training repo: https://github.com/Atten4Vis/LW-DETR.

cc @sbucaille

### CODE REVIEW

File: src/transformers/models/lw_detr/modular_lw_detr.py [hunk 1]
```diff
@@ -1260,8 +1260,8 @@ def gen_encoder_output_proposals(self, enc_output, padding_mask, spatial_shapes)
             _cur += height * width
         output_proposals = torch.cat(proposals, 1)
         output_proposals_valid = ((output_proposals > 0.01) & (output_proposals < 0.99)).all(-1, keepdim=True)
-        output_proposals = output_proposals.masked_fill(padding_mask.unsqueeze(-1), float("inf"))
-        output_proposals = output_proposals.masked_fill(~output_proposals_valid, float("inf"))
+        output_proposals = output_proposals.masked_fill(padding_mask.unsqueeze(-1), float(0))
+        output_proposals = output_proposals.masked_fill(~output_proposals_valid, float(0))
```
Comment (yonigozlan) [inline src/transformers/models/lw_detr/modular_lw_detr.py]: I don't think padding with 0s is correct here?

**Review by yonigozlan — COMMENTED** [review]
Hey @NielsRogge ! Happy to merge this if it fixes training, but I don't think the change in padding values is correct here. Could you change that back and see if the training issue is still solved?

File: src/transformers/models/lw_detr/modular_lw_detr.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/lw_detr/modular_lw_detr.py]: Yes, this change was necessary to be consistent with original LW-DETR tiny training.

The original code is [conditional](https://github.com/Atten4Vis/LW-DETR/blob/d5e6e6c4add2d24dafb965ced8b50163c50b9788/models/transformer.py#L110-L118):
* `unsigmoid=True` uses inf
* `unsigmoid=False` uses 0

The tiny recipe of the original repo sets `--bbox_reparam` (which leads to `unsigmoid=False` [here](https://github.com/Atten4Vis/LW-DETR/blob/d5e6e6c4add2d24dafb965ced8b50163c50b9788/models/transformer.py#L226)), see [lwdetr_tiny_coco_train.sh](https://github.com/Atten4Vis/LW-DETR/blob/main/scripts/lwdetr_tiny_coco_train.sh#L28).

It also turns out that "inf" proposals produce NaN boxes during training.

**Comment by sbucaille** [comment]
@NielsRogge thanks for catching these 😅

File: src/transformers/models/lw_detr/modular_lw_detr.py (hunk 1 above)
Comment (yonigozlan) [reply src/transformers/models/lw_detr/modular_lw_detr.py]: In the end it shouldn't matter really, but if the nan appear during training, that means padding boxes are selected as top k during training, which shouldn't happen. Looking at the code (both original and transformers), it does seem like the scores (`group_enc_outputs_class`) corresponding to the padded boxes, are not being padded themselves. Padding them with -inf should solve the issue and maybe improve fine-tuning.
I'll push a fix to this branch, could you then check if finetuning still works as well?

Comment (NielsRogge) [reply src/transformers/models/lw_detr/modular_lw_detr.py]: Yes I can confirm training still works.

**Review by yonigozlan — APPROVED** [review]
Thanks for fixing @NielsRogge , LGTM!
